src/scaraplate/strategies.py

```diff
@@ -305,12 +326,11 @@ def comment_contents(self) -> str:
         else:
             comment_lines.append(f"From {self.template_meta.commit_url}")
 
-        return "".join(f"{line_comment_start} {line}\n" for line in comment_lines)
+        return [f"{line_comment_start} {line}" for line in comment_lines]
 
-    def render_comment(self, comment_contents: str) -> str:
-        comment_lines = comment_contents.split("\n")
+    def render_comment(self, comment_lines: Sequence[str], *, newline: bytes) -> bytes:
         comment_lines = [self._maybe_add_linter_ignore(line) for line in comment_lines]
-        return "\n".join(comment_lines)
+        return b"".join(line.encode("ascii") + newline for line in comment_lines)
```

Comment:
`newline.join(line.encode("ascii") for line in comment_lines)` ?

Reply:
No. Previously `comment_contents` ended with a newline, thus `split` produced a zero-length last string, so a join by `\n` produced a newline-trailed string.

Now this function accepts a list of comment lines, and that join by `\n` would have produced a string without a trailing newline.

This is covered by tests: the `out` strings in yamls all end with a newline.

Reply:
Can we have some common way to handle ending newline no matter of what strategy generates the content? In this case, we case safely join here while in the end when we write a file ending newline will be ensured. I think it's not good to care about such a thing in multiple places there around.

Reply:
I guess we may introduce a new function which accepts a list of lines and a `newline` argument and produce a string which is basically generated as `b"".join(line + newline for line in comment_lines)`. 

But I would prefer to avoid introducing such too lightweight functions (unless they protect against common errors), because it would make strategies code more coupled with internal scaraplate utilities thus making it harder for people to implement their own external strategies. By that I mean that for implementing custom strategies others would probably look at the existing scaraplate strategies for inspiration, and then they would either be forced to use a lot of internal scaraplate stuff (like that `detect_newline` function which is not a part of public API), or they would have to reimplement them in their code (thus making it harder to implement custom strategies).

That said, the more self-contained strategies implementations are, the easier it is to create other strategies.

And in that case I believe such function wouldn't protect against common errors, especially given that not all strategies process input files line-by-line.

Reply:
> I guess we may introduce a new function which accepts a list of lines and a newline argument and produce a string which is basically generated as

I imagined something else like a function which writes some content to file and ensures that the ending newline is present. No one will call from the strategies since it's low level citizen while we'll be ensured that ending newline is not a problem anymore.

Reply:
Oh, I see. Then the problem is that such function would assume that strategies can only return text contents (ascii or utf8), while I suppose we should let them return binary data as well. Distinguishing between the two outside the strategies would make things more complicated.

Reply:
Could be fine to enforce utf-8 in case of string since we still have to write some binary into a file. Text mode implicitly implies some encoding anyway.

Reply:
> Could be fine to enforce utf-8 in case of string since we still have to write some binary into a file.

Binary data I'm telling about should not be decoded/encoded using a text encoding.

> Text mode implicitly implies some encoding anyway.

Sure, but:
1. Binary data returned by strategies should not be assumed to always be in text mode (unless the strategy explicitly states that, and we don't have such a mechanism yet)
2. I believe it should be handled solely by the strategies, to give the custom strategies all control over the contents of the generated files (and that's why we always pass `BinaryIO` instead of `TextIO`).

Reply:
All true. In case of binary return we can do nothing. In case of text return encode as utf-8. In case of IO return...well, still need some if-else case to handle this type of things.  Looks like we still forced to have some switch about how to handle data type to write a file and there ensure about ending newline is not a problem and actually a sweet spot for that.

Reply:
Let me sum this up. I see 2 options:

1. Let the strategies return texts, and do the text encoding + newlines normalization outside the strategies (this is what you're suggesting so far, if my understanding is correct).
2. Let the strategies do whatever they want, and always treat their result as binary data (this is how it is already done in master and in this PR).

As I said before, the first option would give strategies less control over the resulting data, which is a restriction I don't want to introduce.

>  In case of text return encode as utf-8.

If someone would want a strategy to return text in other encoding than utf8, they would be forced to resort to hacky workarounds such as encoding it in the strategy so that the result would be treated as binary (to avoid the implicit encoding to utf8), and the problem we are trying to solve -- the newlines normalization -- will not be solved there, as the strategy will not do it. OTOH, if all the strategies would do the normalization explicitly, it will be easy for others to copy the built-in strategies' implementation and change whatever they want (e.g. the text encoding) while preserving the already existing newlines normalization code.

Or someone might not want that newlines normalization at all. We should let people write whatever they want in their strategies.

> In case of IO return...well, still need some if-else case to handle this type of things. Looks like we still forced to have some switch about how to handle data type to write a file and there ensure about ending newline is not a problem and actually a sweet spot for that.

And that would make things way more complicated than they're now. We don't need any if-elses if we always treat the result as binary data.

--

I would suggest to not change everything over an issue which we are not really sure yet is actually a problem. Let's keep it as is and see how it goes. If we face any regressions, we can revisit this discussion.

Reply:
> If someone would want a strategy to return text in other encoding than utf8

And per well known anecdote he should be called as extremist. 

I'm fine with current PR, just wanted to find a way to handle newline problem in general way since it will be a common issue for most of strategies.

Reply:
I see the intention, I just don't see a better way to achieve that without hurting the ease of extensibility or power of the strategies.

Reply:
:beers: then. Ship it!